### ISSUE TITLE
Missing errors for `}` braces before `else` in `let...else` statement

### ISSUE THREAD

**Issue by Veykril** [rust-lang/rust-analyzer#18345]
The following are snippets that are denied by rustc, but rust-analyzer's parser does not emit an error.
```rs


struct X {a: i32}
fn f() {
    let foo = X {
        a: 1
    } else {
        //~^ ERROR right curly brace `}` before `else` in a `let...else` statement not allowed
        return;
    };
}

fn i() {
    let foo = &{
        1
    } else {
        //~^ ERROR right curly brace `}` before `else` in a `let...else` statement not allowed
        return;
    };
}

fn j() {
    let bar = 0;
    let foo = bar = {
        1
    } else {
        //~^ ERROR right curly brace `}` before `else` in a `let...else` statement not allowed
        return;
    };
}

fn k() {
    let foo = 1 + {
        1
    } else {
        //~^ ERROR right curly brace `}` before `else` in a `let...else` statement not allowed
        return;
    };
}

fn l() {
    let foo = 1..{
        1
    } else {
        //~^ ERROR right curly brace `}` before `else` in a `let...else` statement not allowed
        return;
    };
}

fn m() {
    let foo = return {
        ()
    } else {
        //~^ ERROR right curly brace `}` before `else` in a `let...else` statement not allowed
        return;
    };
}

fn n() {
    let foo = -{
        1
    } else {
        //~^ ERROR right curly brace `}` before `else` in a `let...else` statement not allowed
        return;
    };
}

fn o() -> Result<(), ()> {
    let foo = do yeet {
        ()
    } else {
        //~^ ERROR right curly brace `}` before `else` in a `let...else` statement not allowed
        return Ok(());
    };
}

fn p() {
    let foo = become {
        ()
    } else {
        //~^ ERROR right curly brace `}` before `else` in a `let...else` statement not allowed
        return;
    };
}

fn q() {
    let foo = |x: i32| {
        x
    } else {
        //~^ ERROR right curly brace `}` before `else` in a `let...else` statement not allowed
        return;
    };
}

fn r() {
    let ok = format_args!("") else { return; };

    let bad = format_args! {""} else { return; };
    //~^ ERROR right curly brace `}` before `else` in a `let...else` statement not allowed
}

fn s() {
    macro_rules! a {
        () => { {} }
    }

    macro_rules! b {
        (1) => {
            let x = a!() else { return; };
        };
        (2) => {
            let x = a! {} else { return; };
            //~^ ERROR right curly brace `}` before `else` in a `let...else` statement not allowed
        };
    }

    b!(1); b!(2);
}

```

These are taken from rustc's test suite

**Comment by roife** [comment]
I noticed that rustc implements it by recursively checking `ast::Expr`:

https://github.com/rust-lang/rust/blob/a06b7cbe21967a86050fa92dab843c8afda1c28e/compiler/rustc_ast/src/util/classify.rs#L171-L239

Currently, we’re check it through `SyntaxKind` during parsing, which is less precise:

https://github.com/rust-lang/rust-analyzer/blob/26d2fbaef2d68a2fccf55d725915188c2634439a/crates/parser/src/grammar/expressions.rs#L136-L142

Perhaps it is better to implement the check in `ide-diagnostics` 👀.
